src/main/java/com/deavensoft/paketomat/dispatcher/DispatcherServiceImpl.java

```diff
@@ -1,27 +1,142 @@
-import java.util.Optional;
+import java.io.IOException;
+import java.util.*;
 
+@AllArgsConstructor
 @Service
 public class DispatcherServiceImpl implements DispatcherService {
-    private final DispatcherRepository dispatcherRepository;
+    private DispatcherRepository dispatcherRepository;
+    @Value("${spring.external.api.distance.url}")
+    private String url;
+    @Value("${spring.external.api.distance.header.host.name}")
+    private String hostName;
+    @Value("${spring.external.api.distance.header.host.value}")
+    private String hostValue;
+    @Value("${spring.external.api.distance.header.key.name}")
+    private String keyName;
+    @Value("${spring.external.api.distance.header.key.value}")
+    private String keyValue;
+    private UserService userService;
+    private EmailService emailService;
+
+    private  static final int MIN_POPULATION = 10000;
+    private  static final int SIZE_OF_PAKETOMAT = 5;
+
     @Autowired
-    public DispatcherServiceImpl(@Qualifier("dispatcher") DispatcherRepository dispatcherRepository){
+    public DispatcherServiceImpl(@Qualifier("dispatcher") DispatcherRepository dispatcherRepository,
+                                 UserService userService, EmailService emailService) {
         this.dispatcherRepository = dispatcherRepository;
+        this.userService = userService;
+        this.emailService = emailService;
     }
-    public List<DispatcherModel> findAllDispatchers(){
+
+    public List<Dispatcher> findAllDispatchers() {
         return dispatcherRepository.findAll();
     }
-    public void saveDispatcher(DispatcherModel newDispatcher){
+
+    public void saveDispatcher(Dispatcher newDispatcher) {
         dispatcherRepository.save(newDispatcher);
     }
-    public Optional<DispatcherModel> findDispatcherById(Long id){
-       return  dispatcherRepository.findById(id);
+
+    public Optional<Dispatcher> findDispatcherById(Long id) {
+        return dispatcherRepository.findById(id);
     }
-    public void deleteDispatcherById(Long id){
+
+    public void deleteDispatcherById(Long id) {
         dispatcherRepository.deleteById(id);
     }
+
+    @Override
+    public void delieverPackage(Package newPackage) throws IOException, NoSuchUserException {
+        Optional<User> user = userService.findUserById(newPackage.getUser().getId());
+        if(user.isPresent()){
+            String[] parts = user.get().getAddress().split(",", 2);
+            String cityUser = parts[1];
+            findNearestCity(cityUser, newPackage);
+            sendMailToCourier();
+        }else
+            throw new NoSuchUserException("There is no user with id " + newPackage.getUser().getId(), HttpStatus.NOT_FOUND, 404);
+    }
+
+    public void findNearestCity(String address, Package newPackage) throws IOException {
+        City city = null;
+        double minDistance = 1000.0;
+        double distance;
+        ArrayList<City> cities = new ArrayList<>(Center.cities);
+        while (true) {
```

Comment:
What will be if any city couldn't be found?

Reply:
Added exception